hsm.py

```diff
@@ -0,0 +1,181 @@
+"""
+Implementation of the HSM (hierachrical state machine) or
+NFSM (nested finite state machine) C++ example from
+http://www.eventhelix.com/RealtimeMantra/HierarchicalStateMachine.htm#.VwqLVEL950w
+in Python
+
+- single source 'message type' for state transition changes
+- message type considered, messages (comment) not considered to avoid complexity
+"""
+
+
+class UnsupportedMessageType(BaseException):
+    pass
+
+
+class UnsupportedState(BaseException):
+    pass
+
+
+class UnsupportedTransition(BaseException):
+    pass
+
+
+class HierachicalStateMachine(object):
+
+    def __init__(self):
+        self._active_state = Active(self)  # Unit.Inservice.Active()
+        self._standby_state = Standby(self)  # Unit.Inservice.Standby()
+        self._suspect_state = Suspect(self)  # Unit.OutOfService.Suspect()
+        self._failed_state = Failed(self)  # Unit.OutOfService.Failed()
+        self._current_state = self._standby_state
+        self.states = {'active': self._active_state,
+                       'standby': self._standby_state,
+                       'suspect': self._suspect_state,
+                       'failed': self._failed_state}
+        self.message_types = {'fault trigger': self._current_state.on_fault_trigger,
+                              'switchover': self._current_state.on_switchover,
+                              'diagnostics passed': self._current_state.on_diagnostics_passed,
+                              'diagnostics failed': self._current_state.on_diagnostics_failed,
+                              'operator inservice': self._current_state.on_operator_inservice}
+
+    def _next_state(self, state):
+        if state in self.states.keys():
```

Comment:
Python uses [EAFP](https://docs.python.org/2/glossary.html#term-eafp) coding style.

> Easier to ask for forgiveness than permission.

So, you code would look like this.

       
```python
try:
    self._current_state = self.states[state]
except KeyError:
    raise UnsupportedState
```